helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -290,7 +327,151 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
     } // end of for-each-partition
   }
 
-  private boolean shouldCreateSTCancellation(Message pendingMessage, String desiredState,
+  /**
+   * Calculate the current active replica count based on state model type.
+   * This method determines how many replicas are currently serving traffic for a partition by
+   * analyzing the current state distribution and applying state model-specific rules. The count includes
+   * replicas in top states, secondary top states (where applicable), and ERROR states since helix considers
+   * them active.
+   * State model handling:
+   * - Single-top state models: Differentiates between patterns with and without secondary top
+   * states
+   * - ONLINE-OFFLINE: Counts ONLINE + ERROR states only
+   * - MASTER-SLAVE-OFFLINE: Counts MASTER + SLAVE + ERROR states
+   * - ONLINE-STANDBY-OFFLINE: Counts ONLINE + STANDBY + ERROR states
+   * - Multi-top state models: Counts only top states + ERROR states
+   * @param currentStateMap Map of instance name to current state for this partition, representing
+   *          the actual state of each replica before any pending transitions
+   * @param stateModelDef State model definition containing the state hierarchy and transition rules
+   *          used to determine which states are considered active
+   * @return The number of replicas currently in active states, used to determine the
+   *         currentActiveReplicaNumber for the partition.
+   */
+  private int calculateCurrentActiveReplicaCount(Map<String, String> currentStateMap,
+      StateModelDefinition stateModelDef) {
+    if (stateModelDef.isSingleTopStateModel()) {
+      return calculateSingleTopStateActiveCount(currentStateMap, stateModelDef);
```

Comment:
I think it is too complex here. All we need is, get the secondary top state count, get top state, include error, exclude dropped and offline.